### PR TITLE
River routing for GEOSldas

### PR DESCRIPTION
**Pull request by zyj8881357**
Pull Request Overview

This pull request initiates the implementation of a river routing process within GEOS, starting with an offline model.

Current Implementation: Only the offline version of the river routing model is introduced at this stage.

All model code is located in the directory:
GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp

The model reads runoff data from the Catchment model outputs (variables RUNOFF and BASEFLOW, with type tavg24_2d_lnd_Nx) and generates daily river-related outputs for each catchment area.

For detailed information on configuring and running the offline model, please refer to readme.txt in the above directory.

Next Steps:
This is an initial implementation, and many aspects require further discussion with relevant members to plan the next steps. 

Additionally, work is ongoing to develop a pre-processing package for creating input data for the routing model.

### CODE REVIEW

**Comment by zyj8881357** [comment]
New Updates:

    1. The river routing model (excluding the reservoir and module components) is now integrated into GEOSldas.
    2. Currently, the model supports EASEv2 M09 and M36 grids. To enable the river routing feature, set RUN_ROUTE to 1 in the exeinp file.
    3. Output Details
        River Discharge [m3 s-1]: Daily river discharge data is saved in the "river" directory (created automatically) within the experiment folder.
        The output is in vector format (*.txt files), where each file contains 291,284 values. These values correspond to catchments indexed from 1 to 291,284.
     4. Restart Files: Restart files are generated monthly in the restart folder and can be read automatically as it exists. If there is no restart file. Zero water storage will be used at the start.

     To convert the vector data into spatial maps, use the 1-minute resolution catchment distribution map available in the CatchIndex variable of the SRTM_PfafData.nc file.
Path: /discover/nobackup/projects/gmao/bcs_shared/make_bcs_inputs/land/topo/v1/SRTM-TopoData/SRTM_PfafData.nc

    For questions or further assistance, please contact Yujin Zeng at:
    Email: yujin.zeng@nasa.gov

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/reservoir.F90 [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+  ! Mark active reservoirs based on type or flood control status
+  do i = 1, nc
+    if(type_res(i) /= 0 .or. fld_res(i) == 1) then
+      active_res(i) = 1
+    endif
+  enddo
+
+  ! Deactivate reservoirs if the use_res flag is set to False
+  if(use_res == .False.) active_res = 0
```
Comment (mathomp4) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/reservoir.F90:258]:
```suggestion
  if(use_res .eqv. .False.) active_res = 0
```

GNU was tripping on this as, per Fortran, you have to use `.eqv.` and `.neqv.` to compare logicals.

**Comment by zyj8881357** [comment]
New Updates:

The pre-processing package for the input data to routing model has been added. Related comments and documents were supplemented. 
At this stage, the CMAKE file is not created for the pre-processing, so the package can be built and run separately (no need building together with the GEOSldas). It is easier for anyone who wants to test and run.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/reservoir.F90 (hunk 1 above)
Comment (zyj8881357) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/reservoir.F90]: I just fixed this. Thank you!

Comment (mathomp4) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/reservoir.F90]: @zyj8881357 It looks like GNU is finding a couple more. Lines 1007 and 1057 in `GEOS_RouteGridComp.F90`

I also see similar in lines 66 and 250 in `lake_mod.F90`

Comment (zyj8881357) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/reservoir.F90]: Those have been fixed. Thank you!

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOS_LandGridComp.F90 [hunk 2]
```diff
@@ -31,7 +31,7 @@ module GEOS_LandGridCompMod
   use GEOS_CatchGridCompMod,   only : CatchSetServices    => SetServices
   use GEOS_CatchCNGridCompMod, only : CatchCNSetServices  => SetServices
   use GEOS_IgniGridCompMod,    only : IgniSetServices     => SetServices
-!  use GEOS_RouteGridCompMod,   only : RouteSetServices    => SetServices
+  use GEOS_RouteGridCompMod,   only : RouteSetServices    => SetServices
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOS_LandGridComp.F90:34]: This comment is not about this particular line, but rather about the placement of GEOS_RouteGridComp inside GEOS_LandGridComp.  Ultimately, we will need to route runoff from the land and the landice tiles.  To do that, I think we need to place GEOS_RouteGridComp one level up, within GEOS_SurfaceGridComp (that is, next to GEOS_LandGridComp).  Making this change will likely require extensive re-wiring.  I wonder if it's best to tackle this re-wiring first or delay the move until later.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 3]
```diff
@@ -18,39 +18,97 @@ module GEOS_RouteGridCompMod
...
+  use reservoir,              ONLY: res_init, res_cal
+  use catch_constants,        ONLY: N_CatG => CATCH_N_PFAFS
+
+  use, intrinsic :: iso_c_binding
   
   implicit none
-  integer, parameter :: N_CatG = 291284
+
+  integer, parameter :: upmax    = 34
+  character(len=500) :: inputdir = "/discover/nobackup/yzeng3/data/river_input/"
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:36]: We cannot hardwire a user-specific directory here. The inputs that come from this directory need to become part of the bcs datasets.  See also comment below about adding parameters into routing restart files.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 4]
```diff
@@ -18,39 +18,97 @@ module GEOS_RouteGridCompMod
...
+  logical, parameter :: use_res  = .True.
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:37]: Are we sure that we want to hardwire using the reservoir model (if that's what this switch does)?  It may be better to include config information in the "RUN_ROUTE" resource variable.  E.g., RUN_ROUTE=1 could be the routing model without reservoirs, and RUN_ROUTE=2 could be the routing model with reservoirs

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 5]
```diff
@@ -18,39 +18,97 @@ module GEOS_RouteGridCompMod
...
   use ESMF
   use MAPL_Mod
   use MAPL_ConstantsMod
-  use ROUTING_MODEL,          ONLY:     &
-       river_routing, ROUTE_DT
-#if 0
-  USE catch_constants, ONLY:          &
-       N_CatG => N_Pfaf_Catchs  
-#endif
+  use ROUTING_MODEL,          ONLY: river_routing_lin, river_routing_hyd, ROUTE_DT
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:27]: river_routing_lin() is commented out below.  Do we want to make this a runtime option?  We could use RUN_ROUTE for this purpose.  E.g., RUN_ROUTE=1 runs the hydraulic model, and RUN_ROUTE=2 runs the linear model.  If we never want  to run the linear model, it should be removed from the code.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 6]
```diff
@@ -398,153 +392,299 @@ subroutine INITIALIZE (GC,IMPORT, EXPORT, CLOCK, RC )
...
-    RETURN_(ESMF_SUCCESS)
+    RETURN_(ESMF_SUCCESS)      
   end subroutine INITIALIZE
+
+  ! --------------------------------------------------------------------------------  
   
 ! -----------------------------------------------------------
 ! RUN -- Run method for the route component
 ! -----------------------------------------------------------
+  subroutine RUN1 (GC,IMPORT, EXPORT, CLOCK, RC )
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:671]: Why define an empty subroutine RUN1()?  Do we need this for anything?  If there's a call to RUN1(), it should be removed, then we can remove the empty subroutine.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 7]
```diff
@@ -398,153 +392,299 @@ subroutine INITIALIZE (GC,IMPORT, EXPORT, CLOCK, RC )
...
+    arbSeq_pf = [(i, i = minCatch, maxCatch)]   
+    route%pfaf => arbSeq_pf
+    route%ntiles = ntiles  
     route%minCatch = minCatch
-    route%maxCatch = maxCatch
+    route%maxCatch = maxCatch 
+
+    ! Read sub-catchment data 
+    allocate(nsub_global(N_CatG),subarea_global(nmax,N_CatG))
+    open(77,file=trim(inputdir)//"/Pfaf_nsub_"//trim(resname)//".txt",status="old",action="read"); read(77,*)nsub_global; close(77)
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:439]: Not sure what this and many subsequent read statements do, but my sense is that we need to include the routing model parameters in the routing model restart files.  Then the "init" routine would not need tap into external files.  
Normally, the process for land model parameters is as follows:
1) Create parameter datasets that can be archived in the bcs_shared directory (along with all the other bcs files).  This is accomplished by make_bcs (and may already be part of the present PR).
2) Create a routing restart file that includes the routing model parameters.  This interfaces with the "remap_restarts" package in the GEOS_Util repo. 
3) When the model is run (in GEOSldas or GEOSgcm), the routing model parameters are read in from the restart file.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 8]
```diff
@@ -398,153 +392,299 @@ subroutine INITIALIZE (GC,IMPORT, EXPORT, CLOCK, RC )
...
+    ! Read restart data
+    call ESMF_ClockGet(clock, currTime=CurrentTime, rc=status)
+    call ESMF_TimeGet(CurrentTime, yy=YY, mm=MM, dd=DD, h=HH, m=MMM, s=SS, rc=status) 
+    write(yr_s,'(I4.4)')YY
+    write(mon_s,'(I2.2)')MM
+    write(day_s,'(I2.2)')DD    
+    if(mapl_am_I_root())print *, "init time is ", YY, "/", MM, "/", DD, " ", HH, ":", MMM, ":", SS    
+    allocate(wriver(ntiles),wstream(ntiles),wres(ntiles))
+    allocate(wriver_global(n_catg),wstream_global(n_catg),wres_global(n_catg))
+    open(77,file="../input/restart/river_storage_rs_"//trim(yr_s)//trim(mon_s)//trim(day_s)//".txt",status="old",action="read",iostat=status)
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:533]: We should not introduce custom, ASCII restart files for the routing model.  Instead, we need to make sure the restarts are handled by MAPL and are netcdf-formatted.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 9]
```diff
@@ -615,502 +757,462 @@ subroutine RUN (GC,IMPORT, EXPORT, CLOCK, RC )
...
+               qoutflow_global, route%scounts_cat, route%rdispls_cat,MPI_REAL, &
+               MPI_COMM_WORLD, mpierr)        
+          !call MPI_allgatherv  (                          &
+          !     route%qsflow_acc,  route%scounts_cat(mype+1)      ,MPI_REAL, &
+          !     qsflow_global, route%scounts_cat, route%rdispls_cat,MPI_REAL, &
+          !     MPI_COMM_WORLD, mpierr)      
+          if(mapl_am_I_root())then   
+             !open(88,file="../river/river_storage_"//trim(yr_s)//trim(mon_s)//trim(day_s)//".txt",action="write")
+             !open(89,file="../river/stream_storage_"//trim(yr_s)//trim(mon_s)//trim(day_s)//".txt",action="write")
+             open(90,file="../river/river_flow_"//trim(yr_s)//trim(mon_s)//trim(day_s)//".txt",action="write")
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:990]: Output from the routing model should be handled by MAPL HISTORY.  We can't publish custom ASCII files with GEOS products.

**Review by gmao-rreichle — COMMENTED** [review]
@zyj8881357: I started reviewing the routing model PR.  I added a couple of minor commits with some cleanup (just a start).  More importantly, we'll need to address the comments below, most of which are quite major and will require substantial revisions in how the model is integrated in GEOS.  
cc: @weiyuan-jiang @tclune @biljanaorescanin @rdkoster @amolod @lcandre2

**Comment by weiyuan-jiang** [comment]
I am quite confused about the variable names.  "N_catg" was used as global tile number before. Can it be changed to N_pfaf_g  here?  The "ntiles" here seem n_pfaf_local to me. ( correct me if I am wrong)

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 10]
```diff
@@ -615,502 +761,462 @@ subroutine RUN (GC,IMPORT, EXPORT, CLOCK, RC )
...
+       if(FirstTime.and.mapl_am_I_root()) print *,"nmax=",nmax
+       allocate(RUNOFF_ACT(ntiles))
+       RUNOFF_ACT=0.
+       do i=1,ntiles
+          do j=1,nmax
+             it=route%subi(j,i) 
+             if(it>0)then
+                RUNOFF_ACT(i)=RUNOFF_ACT(i)+route%subarea(j,i)*runoff_global(it)/1000.   
+             endif
+             if(it==0)exit
```
Comment (weiyuan-jiang) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:884]: exit or cycle ?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 11]
```diff
@@ -615,502 +761,462 @@ subroutine RUN (GC,IMPORT, EXPORT, CLOCK, RC )
...
+       enddo
+       QOUT_CAT = QOUTFLOW_ACT              
+       where(res%active_res==1) QOUT_CAT=QRES_ACT
+
+       ! Collects dishcarge (routing model output) from all processors
+       allocate(QOUTFLOW_GLOBAL(n_catg))
+       call MPI_allgatherv  (                          &
+            QOUT_CAT,  route%scounts_cat(mype+1)      ,MPI_REAL, &
+            QOUTFLOW_GLOBAL, route%scounts_cat, route%rdispls_cat,MPI_REAL, &
+            MPI_COMM_WORLD, mpierr)
```
Comment (weiyuan-jiang) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:931]: All the MPI_COMM_WORLD should be replaced be the comm obtained from VM

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 12]
```diff
@@ -398,153 +393,302 @@ subroutine INITIALIZE (GC,IMPORT, EXPORT, CLOCK, RC )
...
-    call MAPL_LocStreamGet(locstream, tilei=pfaf, OnAttachedGrid=.false., &
+    VERIFY_(STATUS) 
+    ! extract Pfaf (TILEI on the "other" grid)    
+    call MAPL_LocStreamGet(locstream, &
          tileGrid=tilegrid, nt_global=nt_global, RC=status)
-    VERIFY_(STATUS)
-    
+    VERIFY_(STATUS)     
+    route%nt_global = nt_global
+    ! Determine the resolution
```
Comment (weiyuan-jiang) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:412]: use gridname to determine the resolution

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 [hunk 13]
```diff
@@ -398,153 +393,302 @@ subroutine INITIALIZE (GC,IMPORT, EXPORT, CLOCK, RC )
     route%comm = comm
     route%ndes = ndes
     route%mype = mype
-    
-    ! define minCatch, maxCatch
+ 
+    allocate(ims(1:ndes))
+    ! define catchment space for this processor
     call MAPL_DecomposeDim ( n_catg,ims,ndes ) ! ims(mype+1) gives the size of my partition
```
Comment (weiyuan-jiang) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:399]: It seems to to me the distribution of pfaf has nothing to do with the distribution of tiles.  Is there any proximity that can be taken advantage of ?

**Comment by zyj8881357** [comment]
> I am quite confused about the variable names. "N_catg" was used as global tile number before. Can it be changed to N_pfaf_g here? The "ntiles" here seem n_pfaf_local to me. ( correct me if I am wrong)

I changed N_catg to N_pfaf_g through the script.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 (hunk 10 above)
Comment (zyj8881357) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:892]: Thanks for the reminder! As I checked, this should be 'exit'.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 (hunk 12 above)
Comment (zyj8881357) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90:420]: Changed based on this comment.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90 (hunk 13 above)
Comment (zyj8881357) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90]: Since I have no knowledge of the tile distribution, the code was written assuming that the distribution of Pfaf catchments is independent of the tile distribution. I am not sure if there is any proximity that can be taken advantage of.

Comment (gmao-rreichle) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOSroute_GridComp/GEOS_RouteGridComp.F90]: By design, each Pfafstetter catchment can be decomposed into a set of tiles.  Each tile knows the Pfafstetter catchment to which it belongs, and each tile belongs to only one Pfafstetter catchment.  Viewed differently, the tiles are the intersection of the Pfafstetter catchments and the atmospheric grid (cube-sphere or EASE). 

In the LDAS, this is encoded in the "tile_coord" structure, which is assembled from the tile file (*.til): 
https://github.com/GEOS-ESM/GEOSldas_GridComp/blob/f1e382c4b4983e46130d59120f3854d3996df917/LDAS_Shared/LDAS_TileCoordType.F90#L53
There is also information on how much of the Catchment the tile occupies:
https://github.com/GEOS-ESM/GEOSldas_GridComp/blob/f1e382c4b4983e46130d59120f3854d3996df917/LDAS_Shared/LDAS_TileCoordType.F90#L68
If you pick a Pfafstetter catchment and collect all the tiles that belong to it (via "pfaf_index"), then the sum of "frac_pfaf" across these tiles should be 1.   In the interest of full disclosure, I haven't verified this. 

As mentioned above, the tile_coord structure just another form of the information in the tile file (*.til).  This information should also be built into the MAPL "LocationStream" information, but I have no idea if/how this works.  My gut feeling is that "LocationStream" doesn't know this stuff.  However, since "LocationStream" must know how the tiles relate to the atm grid cells, the general infrastructure already exists for mapping from tile space to atm grid space (based on i_indg, j_indg, and frac_cell), and the same kind of operation is needed to map from tile space to Pfafstetter catchment space (based on pfaf_index and frac_pfaf).  

For the cube-sphere tile space, the above should be fairly obvious from inspection of the tile_coord file (in the GEOSldas output).

For the EASE grid, the additional constraint that each EASE grid cell can contain at most 1 tile makes the correspondence between tiles and Pfafstetter catchments a bit weird.  Chances are that there are lots of Pfafstetter catchments that don't have any tiles.  But the pfaf_index and frac_pfaf info should still be there. 
In summary, the tiles and Pfafstetter catchments are very tightly connected.  But please verify what I wrote above. It wouldn't be the first time that the legacy make_bcs processing is in error.

**Comment by gmao-rreichle** [comment]
Closing this PR.  It is superseded by #1143
